cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationManager.java

```diff
@@ -100,19 +99,19 @@ public LocalApplicationManager(CConfiguration configuration, PipelineFactory pip
   }
 
   @Override
-  public ListenableFuture<O> deploy(Id.Namespace id, @Nullable String appId, I input) throws Exception {
+  public ListenableFuture<O> deploy(Id.Namespace namespace, @Nullable String appId, I input) throws Exception {
     Pipeline<O> pipeline = pipelineFactory.getPipeline();
-    pipeline.addLast(new LocalArchiveLoaderStage(store, configuration, id, appId));
+    pipeline.addLast(new LocalArchiveLoaderStage(store, configuration, namespace, appId));
     pipeline.addLast(new ApplicationVerificationStage(store, datasetFramework, adapterService));
-    pipeline.addLast(new DeployDatasetModulesStage(configuration, datasetFramework, inMemoryDatasetFramework));
-    pipeline.addLast(new CreateDatasetInstancesStage(configuration, datasetFramework));
-    pipeline.addLast(new CreateStreamsStage(id, streamAdmin, exploreFacade, exploreEnabled));
+    pipeline.addLast(new DeployDatasetModulesStage(configuration, namespace, datasetFramework,
+                                                   inMemoryDatasetFramework));
+    pipeline.addLast(new CreateDatasetInstancesStage(configuration, datasetFramework, namespace));
+    pipeline.addLast(new CreateStreamsStage(namespace, streamAdmin, exploreFacade, exploreEnabled));
     pipeline.addLast(new DeletedProgramHandlerStage(store, programTerminator, streamConsumerFactory,
                                                     queueAdmin, discoveryServiceClient));
     pipeline.addLast(new ProgramGenerationStage(configuration, namespacedLocationFactory));
     pipeline.addLast(new ApplicationRegistrationStage(store));
-    pipeline.addLast(new CreateSchedulesStage(scheduler));
-    pipeline.setFinally(new DeployCleanupStage());
```

Comment:
this was performing cleanup at the end. I changed it so the stage performs clean up on itself so that it is simpler and less error prone. For example, before, deploy dataset modules could unpack some jars, then you could have have an error in a later stage and those jars might not get cleaned up.